sdks/java/core/src/main/java/org/apache/beam/sdk/coders/ZstdCoder.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.coders;
+
+import com.github.luben.zstd.Zstd;
+import com.github.luben.zstd.ZstdCompressCtx;
+import com.github.luben.zstd.ZstdDecompressCtx;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.util.CoderUtils;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+/**
+ * Wraps an existing coder with Zstandard compression. It makes sense to use this coder when it's
+ * likely that the encoded value is quite large and compressible or when a dictionary is available
+ * to improve compression performance.
+ */
+public class ZstdCoder<T> extends StructuredCoder<T> {
+  private final Coder<T> innerCoder;
+  private final @Nullable byte[] dict;
+  private final int level;
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, byte[] dict, int level) {
+    return new ZstdCoder<>(innerCoder, dict, level);
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, byte[] dict) {
+    return new ZstdCoder<>(innerCoder, dict, Zstd.defaultCompressionLevel());
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder, int level) {
+    return new ZstdCoder<>(innerCoder, null, level);
+  }
+
+  /** Wraps the given coder into a {@link ZstdCoder}. */
+  public static <T> ZstdCoder<T> of(Coder<T> innerCoder) {
+    return new ZstdCoder<>(innerCoder, null, Zstd.defaultCompressionLevel());
+  }
+
+  private ZstdCoder(Coder<T> innerCoder, @Nullable byte[] dict, int level) {
+    this.innerCoder = innerCoder;
+    this.dict = dict;
+    this.level = level;
+  }
+
+  @Override
+  public void encode(T value, OutputStream os) throws IOException {
+    ZstdCompressCtx ctx = new ZstdCompressCtx();
```

Comment:
creating and destroying contexts is actually pretty expensive, sometimes more so than the actual compression operation, you should try to reuse these as much as possible.  A ThreadLocal could be useful in this case.

Reply:
Note that there hasn't been a guarantee that the same coder instance isn't reused elsewhere and hence needs to be re-entrant (e.g. `ZstdCoder(ZstdCoder(StringCoder))`.

It would be nice to relax this condition in Beam.

Reply:
yeah that'd be nice.  imo the most useful guarantee would be that coders are never shared between threads concurrently (like DoFns) but that might be a more complicated problem to solve.

Reply:
I initially tried to address this by using a member variable of type EmptyOnDeserializationThreadLocal, as is used in JAXBCoder for example, but this lead to a significant increase in memory usage.
It seems like the context instances aren't closed and collected properly when the thread stops.
Apache Lucene describes this issue and addresses it for their needs with [CloseableThreadLocal](https://lucene.apache.org/core/9_4_1/core/org/apache/lucene/util/CloseableThreadLocal.html).

Reply:
An alternative would be to use an AtomicReference that stores a context object that can be re-used.

e.g.
```
Context context = sharedContext.getAndSet(null);
if (context == null) {
  context = new Context ...
}
... use context to encode ...
sharedContext.lazySet(context);
```